README
This library contains an R implementation of some triangulation routines.  

It is based on Fortran code from R. J. Renka in the ACM Collected 
Algorithms archive under

  http://www.netlib.org/toms/751

  R. J. Renka (1996). Algorithm 751: TRIPACK: a constrained
  two-dimensional {Delaunay} triangulation package.
  ACM Transactions on Mathematical Software.
  22, 1-8.

I added the Fortran files inhull.f and voronoi.f wich implements additional
subroutines for determination of convex hulls and voronoi mosaics.

Currently the library  contains the following R functions which provide access
to the Fortran subroutines of TRIPACK and implements new objects for triangu-
lations and voronoi mosaics:

add.constraint           
convex.hull             
identify.tri           
in.convex.hull        
neighbours          
on.convex.hull     
outer.convhull   
plot.tri       
plot.voronoi  
print.summary.tri        
print.summary.voronoi   
print.tri              
print.tri            
summary.tri         
summary.voronoi    
tri               
tri.find         
tri.mesh        
triangles      
voronoi      
voronoi.mosaic

The help pages are based on the Fortran comments. 

This library was intended by the akima library, which also contains some (but 
not all) of the TRIPACK functions. 

Currently plots of voronoi mosaics may fail (due to numerical problems) if 
the corresponding triangulation contains very small triangles at the boundary.

------------------------------------------------------------------
Albrecht Gebhardt           email: albrecht.gebhardt@uni-klu.ac.at
Institut fuer Mathematik    Tel. : (++43 463) 2700/837
Universitaet Klagenfurt     Fax  : (++43 463) 2700/834
Villacher Str. 161
A-9020 Klagenfurt, Austria
------------------------------------------------------------------

The abstract of the original article at ACM follows:
##############################################################################
                 TRIPACK is a Fortran 77 software package that employs
                 an incremental algorithm to construct a constrained
                 Delaunay triangulation of a set of points in the plane
                 (nodes). The triangulation covers the convex hull of
                 the nodes but may include polygonal constraint regions
                 whose triangles are distinguishable from those in the
                 remainder of the triangulation. This effectively allows
                 for a nonconvex or multiply connected triangulation
                 (the complement of the union of constraint regions)
                 while retaining the efficiency of searching and
                 updating a convex triangulation. The package provides a
                 wide range of capabilities including an efficient means
                 of updating the triangulation with nodal additions or
                 deletions. For N  nodes, the storage requirement is
                 13N integer storage locations in addition to the 2N 
                 nodal coordinates.



ChangeLog
commit 77cbded8e5cfe4207c0aaf783027af629e3bb357
Author: Albrecht Gebhardt <albrecht.gebhardt@aau.at>
Date:   Mon Mar 2 09:08:18 2020 +0100

    change fortran subroutine to int (from void). LTO error otherwise.

commit 0e4865e814b7eb6545afbb624f79f543c83cb348
Author: Albrecht Gebhardt <albrecht.gebhardt@aau.at>
Date:   Fri Feb 28 12:13:01 2020 +0100

    date

commit d2299555a4e039843fe49696d4d722889298df83
Author: Albrecht Gebhardt <albrecht.gebhardt@aau.at>
Date:   Tue Feb 18 10:28:18 2020 +0100

    silence gfortran -flto tests

commit d85770cd928f0adebef009a31395a9b9ca8d2b2d
Author: Albrecht Gebhardt <albrecht.gebhardt@aau.at>
Date:   Mon Feb 17 12:36:08 2020 +0100

    next version

commit 3c4a1ad9fa401281c5490436f6249c2b46eb5c8a
Author: Albrecht Gebhardt <albrecht.gebhardt@aau.at>
Date:   Tue Feb 11 10:33:06 2020 +0100

    try to use R_useDynamicSymbols ...

commit ffd1fe71999b4ff5c25642adb16f53c21577734e
Author: agebhard <>
Date:   Mon Jan 29 18:16:09 2018 +0000

    fix typo

commit c5316f3e2df1998122e2757906df96a2f1c4911f
Author: agebhard <>
Date:   Fri Dec 16 12:15:36 2016 +0000

    * ACM Licence URL updated
    * Some fortran cleanup
    * Added jitter option to triangulation to allow for
      collinear points (in sync with akima package)
    * code under develeopment (additive weigted voronoi) exluded from
      distributed package to avoid confusion.

commit 5978adf8120b9c01f0537a4fb069c2755d3d2576
Author: agebhard <>
Date:   Wed Sep 2 16:17:08 2015 +0000

    unlock functions under development

commit 5e243bd30fb751252e9583f213653e039fcdf0ba
Author: agebhard <>
Date:   Tue Sep 1 17:11:04 2015 +0000

    cleanup for R 3.2.2

commit e9326de066cc8b4efdb35f0271e9d582fa93bdd2
Author: agebhard <>
Date:   Mon Sep 16 08:59:24 2013 +0000

    more excludes

commit c0d8c197fba7da8e8ae616a79d3ca78dad1a7a45
Author: agebhard <>
Date:   Mon Sep 16 08:28:47 2013 +0000

    cleanup

commit ffe6868ae00f1bacb26c317ab9e048feb377d5e7
Author: agebhard <>
Date:   Mon Sep 16 08:25:56 2013 +0000

    cleanup

commit a171ab160713471b68a70a044105a134a8255bb4
Author: agebhard <>
Date:   Mon Sep 16 06:35:43 2013 +0000

    fix license entry

commit 6b5347de2e3555cf3b6278725608c2b1379bd5dc
Author: agebhard <>
Date:   Wed Oct 31 13:02:59 2012 +0000

    wcircum examples

commit cd8e0358fa3e92ac0613c1f248e6a55409c519b0
Author: agebhard <>
Date:   Tue Oct 30 12:08:46 2012 +0000

    remove STOPs and WRITEs

commit bc0f15b10875819ba0320e51cc7ae45c7c952e1e
Author: agebhard <>
Date:   Tue Jan 24 23:13:27 2012 +0000

    more work on wcircum

commit 687a72a895b8ca6cabe316a9bf559455e73f1978
Author: agebhard <>
Date:   Sun Jan 8 22:47:09 2012 +0000

    prepare for R 2.14.1

commit 71b639cb4b09b50b26bc6531a60d9c85cce307c0
Author: agebhard <>
Date:   Fri Mar 4 14:08:05 2011 +0000

    work on wcircum

commit c603485430ffe2a2d975205cd6faf7aa0a69e530
Author: agebhard <>
Date:   Tue Mar 1 21:46:08 2011 +0000

    new function

commit 3e7878d8503ca38f1905ecd5c9594e324bd39d1c
Author: agebhard <>
Date:   Mon Feb 28 14:50:17 2011 +0000

    * include an option for equal scaled plots
    * some documentation fixes

commit d3bff7c9cd70249ee3a2e92b88402a7fb57fdb10
Author: agebhard <>
Date:   Mon Oct 26 19:06:00 2009 +0000

    alias for left

commit c6583bbc634dce710240e67c9fe684b411b41fd9
Author: agebhard <>
Date:   Mon Oct 26 19:05:23 2009 +0000

    circumcircle stuff works now.

commit c9b2cc1836f4f6ad5504b04dbb4c333787ce730f
Author: agebhard <>
Date:   Mon Oct 26 18:24:36 2009 +0000

    test case

commit 82e25c6b831bb2b22ea2c1e9e77241826a33b896
Author: agebhard <>
Date:   Mon Oct 26 18:24:00 2009 +0000

    circumcircle seems to work

commit e73675305b9528897e292a53a5493e25e8a7dede
Author: agebhard <>
Date:   Sun Oct 25 21:59:02 2009 +0000

    more work on circumcircle, still not working.

commit 8639b2e30cabbfb1b00ceb65831101731e978e1c
Author: agebhard <>
Date:   Fri Oct 23 20:20:58 2009 +0000

    add initial version of a circumcircle function for n points.
    needs still some work.

commit 90ed3cfad7a29ea51d59f5001b7afa0f23b70f60
Author: agebhard <>
Date:   Fri Sep 25 09:39:48 2009 +0000

    use Fortran implementaion of LEFT for is.left.of,
    plot.aw routine mostly fixed for simple cases

commit c91759a6663d101aed7ca4084f53a35d51eea3b8
Author: agebhard <>
Date:   Thu Sep 24 18:13:31 2009 +0000

    more aw work

commit 49b89fa02b93c05a9a9f74279bbac96bdc16ba37
Author: agebhard <>
Date:   Thu Sep 24 12:43:56 2009 +0000

    small fix

commit 3ff1c2417a45a15d069a97a4a05209904567666f
Author: agebhard <>
Date:   Thu Sep 24 12:41:57 2009 +0000

    some more aw work.

commit 569197b2b965f8bca6152f1ab2e86089ce937dc1
Author: agebhard <>
Date:   Wed Sep 23 12:57:22 2009 +0000

    maybe helpful in aw.voronoi...

commit a8ea9ba89326d0e1fa74fc6a4a7fcd1aac938816
Author: agebhard <>
Date:   Wed Sep 23 12:56:55 2009 +0000

    still trying to fix problems in additive weighted voronoi code.

commit 6b13bd6163033a01992d394fa085c29f1b9f0046
Author: agebhard <>
Date:   Tue Sep 22 10:25:52 2009 +0000

    somehow complete the aw* functions

commit 48ec4a9932e3d6206b98e6ee924a2eca59c55542
Author: agebhard <>
Date:   Mon Sep 21 09:14:31 2009 +0000

    fixes to pass R CMD check version 2.10.0

commit 27c7a523fce27babaa6085c97667fec25da8aaa6
Author: agebhard <>
Date:   Mon Aug 24 06:26:09 2009 +0000

    small fixes for public release

commit bbfeb41897f2cce3dceb39247fad74aa57555d86
Author: agebhard <>
Date:   Mon Jun 15 11:08:29 2009 +0000

    fixes for R-2.9.0

commit beb93089bacc5abb0ebdd4ba34391621518ae105
Author: agebhard <>
Date:   Wed Jun 10 18:48:36 2009 +0000

    recovered from temp. file deletion

commit 6afdb98c79c884a6ffaa36b5572caf1a408c4474
Author: agebhard <>
Date:   Wed Jun 10 17:54:00 2009 +0000

    License clarification

commit 31021f8d4371e93c9c119fc514eaf9ab5ae2de51
Author: agebhard <>
Date:   Thu Mar 5 07:38:15 2009 +0000

    sort out LICENSE mismatches

commit 80947edf0727538a6980c41572eda03735c4ddb1

src/func.f
*DECK DASINH
      DOUBLE PRECISION FUNCTION DASINH (X)
C***BEGIN PROLOGUE  DASINH
C***PURPOSE  Compute the arc hyperbolic sine.
C***LIBRARY   SLATEC (FNLIB)
C***CATEGORY  C4C
C***TYPE      DOUBLE PRECISION (ASINH-S, DASINH-D, CASINH-C)
C***KEYWORDS  ARC HYPERBOLIC SINE, ASINH, ELEMENTARY FUNCTIONS, FNLIB,
C             INVERSE HYPERBOLIC SINE
C***AUTHOR  Fullerton, W., (LANL)
C***DESCRIPTION
C
C DASINH(X) calculates the double precision arc hyperbolic
C sine for double precision argument X.
C
C***REFERENCES  (NONE)
C***ROUTINES CALLED  D1MACH, DCSEVL, INITDS
C***REVISION HISTORY  (YYMMDD)
C   770601  DATE WRITTEN
C   890531  Changed all specific intrinsics to generic.  (WRB)
C   890531  REVISION DATE from Version 3.2
C   891214  Prologue converted to Version 4.0 format.  (BAB)
C***END PROLOGUE  DASINH
      DOUBLE PRECISION X, ASNHCS(39), ALN2, SQEPS, XMAX, Y,
     1  DCSEVL, D1MACH
      LOGICAL FIRST
      SAVE ASNHCS, ALN2, NTERMS, XMAX, SQEPS, FIRST
      DATA ASNHCS(  1) / -.1282003991 1738186343 3721273592 68 D+0     /
      DATA ASNHCS(  2) / -.5881176118 9951767565 2117571383 62 D-1     /
      DATA ASNHCS(  3) / +.4727465432 2124815640 7252497560 29 D-2     /
      DATA ASNHCS(  4) / -.4938363162 6536172101 3601747902 73 D-3     /
      DATA ASNHCS(  5) / +.5850620705 8557412287 4948352593 21 D-4     /
      DATA ASNHCS(  6) / -.7466998328 9313681354 7550692171 88 D-5     /
      DATA ASNHCS(  7) / +.1001169358 3558199265 9661920158 12 D-5     /
      DATA ASNHCS(  8) / -.1390354385 8708333608 6164722588 86 D-6     /
      DATA ASNHCS(  9) / +.1982316948 3172793547 3173602371 48 D-7     /
      DATA ASNHCS( 10) / -.2884746841 7848843612 7472728003 17 D-8     /
      DATA ASNHCS( 11) / +.4267296546 7159937953 4575149959 07 D-9     /
      DATA ASNHCS( 12) / -.6397608465 4366357868 7526323096 81 D-10    /
      DATA ASNHCS( 13) / +.9699168608 9064704147 8782931311 79 D-11    /
      DATA ASNHCS( 14) / -.1484427697 2043770830 2466583656 96 D-11    /
      DATA ASNHCS( 15) / +.2290373793 9027447988 0401843789 83 D-12    /
      DATA ASNHCS( 16) / -.3558839513 2732645159 9789426513 10 D-13    /
      DATA ASNHCS( 17) / +.5563969408 0056789953 3745390885 54 D-14    /
      DATA ASNHCS( 18) / -.8746250959 9624678045 6665935201 62 D-15    /
      DATA ASNHCS( 19) / +.1381524884 4526692155 8688022981 29 D-15    /
      DATA ASNHCS( 20) / -.2191668828 2900363984 9551422641 49 D-16    /
      DATA ASNHCS( 21) / +.3490465852 4827565638 3139237068 80 D-17    /
      DATA ASNHCS( 22) / -.5578578840 0895742439 6301570321 06 D-18    /
      DATA ASNHCS( 23) / +.8944514661 7134012551 0508827989 33 D-19    /
      DATA ASNHCS( 24) / -.1438342634 6571317305 5518452394 66 D-19    /
      DATA ASNHCS( 25) / +.2319181187 2169963036 3261446826 66 D-20    /
      DATA ASNHCS( 26) / -.3748700795 3314343674 5706045439 99 D-21    /
      DATA ASNHCS( 27) / +.6073210982 2064279404 5492428800 00 D-22    /
      DATA ASNHCS( 28) / -.9859940276 4633583177 3701734400 00 D-23    /
      DATA ASNHCS( 29) / +.1603921745 2788496315 2326382933 33 D-23    /
      DATA ASNHCS( 30) / -.2613884735 0287686596 7161343999 99 D-24    /
      DATA ASNHCS( 31) / +.4267084960 6857390833 3581653333 33 D-25    /
      DATA ASNHCS( 32) / -.6977021703 9185243299 7307733333 33 D-26    /
      DATA ASNHCS( 33) / +.1142508833 6806858659 8126933333 33 D-26    /
      DATA ASNHCS( 34) / -.1873529207 8860968933 0210133333 33 D-27    /
      DATA ASNHCS( 35) / +.3076358441 4464922794 0659200000 00 D-28    /
      DATA ASNHCS( 36) / -.5057736403 1639824787 0463999999 99 D-29    /
      DATA ASNHCS( 37) / +.8325075471 2689142224 2133333333 33 D-30    /
      DATA ASNHCS( 38) / -.1371845728 2501044163 9253333333 33 D-30    /
      DATA ASNHCS( 39) / +.2262986842 6552784104 1066666666 66 D-31    /
      DATA ALN2 / 0.6931471805 5994530941 7232121458 18D0 /
      DATA FIRST /.TRUE./
C***FIRST EXECUTABLE STATEMENT  DASINH
      IF (FIRST) THEN
         NTERMS = INITDS (ASNHCS, 39, 0.1*REAL(D1MACH(3)) )
         SQEPS = SQRT(D1MACH(3))
         XMAX = 1.0D0/SQEPS
      ENDIF
      FIRST = .FALSE.
C
      Y = ABS(X)
      IF (Y.GT.1.0D0) GO TO 20
C
      DASINH = X
      IF (Y.GT.SQEPS) DASINH = X*(1.0D0 + DCSEVL (2.D0*X*X-1.D0,
     1  ASNHCS, NTERMS) )
      RETURN
 20   IF (Y.LT.XMAX) DASINH = LOG (Y+SQRT(Y*Y+1.D0))
      IF (Y.GE.XMAX) DASINH = ALN2 + LOG(Y)
      DASINH = SIGN (DASINH, X)
      RETURN
C
      END
*DECK DCSEVL
      DOUBLE PRECISION FUNCTION DCSEVL (X, CS, N)
C***BEGIN PROLOGUE  DCSEVL
C***PURPOSE  Evaluate a Chebyshev series.
C***LIBRARY   SLATEC (FNLIB)
C***CATEGORY  C3A2
C***TYPE      DOUBLE PRECISION (CSEVL-S, DCSEVL-D)
C***KEYWORDS  CHEBYSHEV SERIES, FNLIB, SPECIAL FUNCTIONS
C***AUTHOR  Fullerton, W., (LANL)
C***DESCRIPTION
C
C  Evaluate the N-term Chebyshev series CS at X.  Adapted from
C  a method presented in the paper by Broucke referenced below.
C
C       Input Arguments --
C  X    value at which the series is to be evaluated.
C  CS   array of N terms of a Chebyshev series.  In evaluating
C       CS, only half the first coefficient is summed.
C  N    number of terms in array CS.
C
C***REFERENCES  R. Broucke, Ten subroutines for the manipulation of
C                 Chebyshev series, Algorithm 446, Communications of
C                 the A.C.M. 16, (1973) pp. 254-256.
C               L. Fox and I. B. Parker, Chebyshev Polynomials in
C                 Numerical Analysis, Oxford University Press, 1968,
C                 page 56.
C***ROUTINES CALLED  D1MACH, XERMSG
C***REVISION HISTORY  (YYMMDD)
C   770401  DATE WRITTEN
C   890831  Modified array declarations.  (WRB)
C   890831  REVISION DATE from Version 3.2
C   891214  Prologue converted to Version 4.0 format.  (BAB)
C   900315  CALLs to XERROR changed to CALLs to XERMSG.  (THJ)
C   900329  Prologued revised extensively and code rewritten to allow
C           X to be slightly outside interval (-1,+1).  (WRB)
C   920501  Reformatted the REFERENCES section.  (WRB)
C***END PROLOGUE  DCSEVL
      DOUBLE PRECISION B0, B1, B2, CS(*), ONEPL, TWOX, X, D1MACH
      LOGICAL FIRST
      SAVE FIRST, ONEPL
      DATA FIRST /.TRUE./
C***FIRST EXECUTABLE STATEMENT  DCSEVL
      IF (FIRST) ONEPL = 1.0D0 + D1MACH(4)
      FIRST = .FALSE.
      IF (N .LT. 1) CALL XERMSG ('SLATEC', 'DCSEVL',
     +   'NUMBER OF TERMS .LE. 0', 2, 2)
      IF (N .GT. 1000) CALL XERMSG ('SLATEC', 'DCSEVL',
     +   'NUMBER OF TERMS .GT. 1000', 3, 2)
      IF (ABS(X) .GT. ONEPL) CALL XERMSG ('SLATEC', 'DCSEVL',
     +   'X OUTSIDE THE INTERVAL (-1,+1)', 1, 1)
C
      B1 = 0.0D0
      B0 = 0.0D0
      TWOX = 2.0D0*X
      DO 10 I = 1,N
         B2 = B1
         B1 = B0
         NI = N + 1 - I
         B0 = TWOX*B1 - B2 + CS(NI)
   10 CONTINUE
C
      DCSEVL = 0.5D0*(B0-B2)
C
      RETURN
      END
*DECK INITDS
      FUNCTION INITDS (OS, NOS, ETA)
C***BEGIN PROLOGUE  INITDS
C***PURPOSE  Determine the number of terms needed in an orthogonal
C            polynomial series so that it meets a specified accuracy.
C***LIBRARY   SLATEC (FNLIB)
C***CATEGORY  C3A2
C***TYPE      DOUBLE PRECISION (INITS-S, INITDS-D)
C***KEYWORDS  CHEBYSHEV, FNLIB, INITIALIZE, ORTHOGONAL POLYNOMIAL,
C             ORTHOGONAL SERIES, SPECIAL FUNCTIONS
C***AUTHOR  Fullerton, W., (LANL)
C***DESCRIPTION
C
C  Initialize the orthogonal series, represented by the array OS, so
C  that INITDS is the number of terms needed to insure the error is no
C  larger than ETA.  Ordinarily, ETA will be chosen to be one-tenth
C  machine precision.
C
C             Input Arguments --
C   OS     double precision array of NOS coefficients in an orthogonal
C          series.
C   NOS    number of coefficients in OS.
C   ETA    single precision scalar containing requested accuracy of
C          series.
C
C***REFERENCES  (NONE)
C***ROUTINES CALLED  XERMSG
C***REVISION HISTORY  (YYMMDD)
C   770601  DATE WRITTEN
C   890531  Changed all specific intrinsics to generic.  (WRB)
C   890831  Modified array declarations.  (WRB)
C   891115  Modified error message.  (WRB)
C   891115  REVISION DATE from Version 3.2
C   891214  Prologue converted to Version 4.0 format.  (BAB)
C   900315  CALLs to XERROR changed to CALLs to XERMSG.  (THJ)
C***END PROLOGUE  INITDS
      DOUBLE PRECISION OS(*)
C***FIRST EXECUTABLE STATEMENT  INITDS
      IF (NOS .LT. 1) CALL XERMSG ('SLATEC', 'INITDS',
     +   'Number of coefficients is less than 1', 2, 1)
C
      ERR = 0.
      DO 10 II = 1,NOS
        I = NOS + 1 - II
        ERR = ERR + ABS(REAL(OS(I)))
        IF (ERR.GT.ETA) GO TO 20
   10 CONTINUE
C
   20 IF (I .EQ. NOS) CALL XERMSG ('SLATEC', 'INITDS',
     +   'Chebyshev series too short for specified accuracy', 1, 1)
      INITDS = I
C
      RETURN
      END
*DECK DACOSH
      DOUBLE PRECISION FUNCTION DACOSH (X)
C***BEGIN PROLOGUE  DACOSH
C***PURPOSE  Compute the arc hyperbolic cosine.
C***LIBRARY   SLATEC (FNLIB)
C***CATEGORY  C4C
C***TYPE      DOUBLE PRECISION (ACOSH-S, DACOSH-D, CACOSH-C)
C***KEYWORDS  ACOSH, ARC HYPERBOLIC COSINE, ELEMENTARY FUNCTIONS, FNLIB,
C             INVERSE HYPERBOLIC COSINE
C***AUTHOR  Fullerton, W., (LANL)
C***DESCRIPTION
C
C DACOSH(X) calculates the double precision arc hyperbolic cosine for
C double precision argument X.  The result is returned on the
C positive branch.
C
C***REFERENCES  (NONE)
C***ROUTINES CALLED  D1MACH, XERMSG
C***REVISION HISTORY  (YYMMDD)
C   770601  DATE WRITTEN
C   890531  Changed all specific intrinsics to generic.  (WRB)
C   890531  REVISION DATE from Version 3.2
C   891214  Prologue converted to Version 4.0 format.  (BAB)
C   900315  CALLs to XERROR changed to CALLs to XERMSG.  (THJ)
C***END PROLOGUE  DACOSH
      DOUBLE PRECISION X, DLN2, XMAX,  D1MACH
      SAVE DLN2, XMAX
      DATA DLN2 / 0.6931471805 5994530941 7232121458 18 D0 /
      DATA XMAX / 0.D0 /
C***FIRST EXECUTABLE STATEMENT  DACOSH
      IF (XMAX.EQ.0.D0) XMAX = 1.0D0/SQRT(D1MACH(3))
C
      IF (X .LT. 1.D0) CALL XERMSG ('SLATEC', 'DACOSH',
     +   'X LESS THAN 1', 1, 2)
C
      IF (X.LT.XMAX) DACOSH = LOG (X+SQRT(X*X-1.0D0))
      IF (X.GE.XMAX) DACOSH = DLN2 + LOG(X)
C
      RETURN
      END
